QC/wdl/tasks/ntsm.wdl

```diff
@@ -85,33 +67,69 @@ task ntsm_count {
     }
     
     parameter_meta {
-        input_fastqs: "Files must be in fastq format"
+        input_reads: "Files must be in fastq, fastq.gz (preferred), BAM, or CRAM format. CRAM files must be passed with cram_reference"
         count_kmer_size: "k-mer size to use (sliding window is applied: highest count is stored)"
     }
 
     # Estimate disk size required
-    Int input_fastq_size  = ceil(size(input_fastqs, "GB"))       
-    Int final_disk_dize   = input_fastq_size * 2 + addldisk
+    Int input_read_size  = ceil(size(input_reads, "GB"))       
+    Int final_disk_dize   = input_read_size * 5 + addldisk
 
     # Create output file name
     String output_counts_fn = "${sample_id}_${read_type}_counts.txt"
 
     command <<<
         set -eux -o pipefail
 
+        READS=(~{sep=" " input_reads})
+
+        ## dump converted reads into output folder
+        mkdir output
+
+        ## convert all reads to fastq.gz format
+        for READFILE in "${READS[@]}"
+            do
+                echo "$READFILE"
+
+                FILENAME="$(basename -- $READFILE)"
+                PREFIX="${FILENAME%.*}"
+                SUFFIX="${FILENAME##*.}"
+
+                if [[ "$SUFFIX" == "bam" ]] ; then
+                    samtools fastq -@~{threadCount} $READFILE > output/${PREFIX}.fastq.gz
+
+                elif [[ "$SUFFIX" == "cram" ]] ; then
+                    if [[ ! -f "~{cram_reference}" ]] ; then
+                        echo "Could not extract $FILENAME, reference file not supplied"
+                        exit 1
+                    fi
+                    ln -s ~{cram_reference}
+                    samtools fastq -@~{threadCount} --reference `basename ~{cram_reference}` $READFILE > output/${PREFIX}.fastq.gz
+                
+                elif [[ "$SUFFIX" == "fastq" ]] ; then
+                    gzip $READFILE > output/${PREFIX}.fastq.gz
+                
+                elif [[ "$SUFFIX" != "gz" ]] ; then
+                    echo "Unsupported file type: ${PREFIX}.${SUFFIX}"
+                    exit 1
+                else
+                    ln $READFILE output/$FILENAME
+                fi
+
+            done
+
+
         ## localize kmers
         ln -s /opt/ntsm/data/31_AT.fa .
         ln -s /opt/ntsm/data/31_CG.fa .
 
-        FASTQS=(~{sep=" " input_fastqs})
-
         ## Count number of k-mers found 
         /opt/ntsm/src/ntsmCount \
             -r 31_AT.fa \
             -a 31_CG.fa \
             -k ~{count_kmer_size} \
             -t ~{threadCount} \
-            <(cat ${FASTQS}) \
```

Comment:
@juklucas Nice job!

I didn't find any issue but here is just a recommendation.
I guess you can save more time by working with `fastq` instead of `fastq.gz`. Since most of the read files we have are in `bam` format so what you are doing now is to extract reads with `samtools fastq` and then compress them with `gzip` and then, while passing to `ntsmCount`, decompress with `gzip -cd` again. You can save time by just saving `fastq` and working with that.

Reply:
Thanks for the recommendation (and the review)! I'll merge, and take a look at speeding it up next time I make changes to the workflow.